# Biodiversity in National Parks - Data Analysis
### Analysis Overview
In this analysis, we will be analyzing data from two datasets containing information on National Parks and the species who inhabit them.

The datasets are:

- ```"observations.csv"``` - includes data on the national park being observed, what species live their, and how many times each species has been recorded being spotted within the last 7 days.
- ```"species_info.csv"``` - includes data on a large number of different species.  It features the species' scientific name, common name, category of species, and conservation status.

---

## Section 1 - Understanding the Data
In order to work with any data, you first need to understand what you'll be working with.  

1. Import the Python libraries you'll need to conduct your analysis 
2. Load the two CSV datasets into two seperate Pandas DataFrames, ```parks``` and ```species```
3. View the first 10 rows of the ```parks``` dataframe
4. View the first 10 rows of the ```species``` dataframe

In [512]:
# 1

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

In [513]:
# 2

parks = pd.read_csv('observations.csv')
species = pd.read_csv('species_info.csv')

In [514]:
# 3 

parks.head(10)

scientific_name                            park_name  \
0                 Vicia benghalensis  Great Smoky Mountains National Park   
1                     Neovison vison  Great Smoky Mountains National Park   
2                  Prunus subcordata               Yosemite National Park   
3               Abutilon theophrasti                  Bryce National Park   
4           Githopsis specularioides  Great Smoky Mountains National Park   
5  Elymus virginicus var. virginicus               Yosemite National Park   
6                   Spizella pusilla            Yellowstone National Park   
7                  Elymus multisetus  Great Smoky Mountains National Park   
8             Lysimachia quadrifolia               Yosemite National Park   
9         Diphyscium cumberlandianum            Yellowstone National Park   

   observations  
0            68  
1            77  
2           138  
3            84  
4            85  
5           112  
6           228  
7            39  
8           168  
9           250

In [515]:
# 4

species.head(10)

category                scientific_name  \
0   Mammal  Clethrionomys gapperi gapperi   
1   Mammal                      Bos bison   
2   Mammal                     Bos taurus   
3   Mammal                     Ovis aries   
4   Mammal                 Cervus elaphus   
5   Mammal         Odocoileus virginianus   
6   Mammal                     Sus scrofa   
7   Mammal                  Canis latrans   
8   Mammal                    Canis lupus   
9   Mammal                    Canis rufus   

                                        common_names conservation_status  
0                           Gapper's Red-Backed Vole                 NaN  
1                              American Bison, Bison                 NaN  
2  Aurochs, Aurochs, Domestic Cattle (Feral), Dom...                 NaN  
3  Domestic Sheep, Mouflon, Red Sheep, Sheep (Feral)                 NaN  
4                                      Wapiti Or Elk                 NaN  
5                                  White-Tailed Deer                 NaN  
6                                Feral Hog, Wild Pig                 NaN  
7                                             Coyote  Species of Concern  
8                                          Gray Wolf          Endangered  
9                                           Red Wolf          Endangered

The first thing that stands out to me is in the ```species``` dataframe.  Without looking any deeper you can already see that it contains some missing values ('NaN').  In order to accurately work with this data, you'll need to clean up your dataframes.  Aside from that one column in the ```species``` dataframe, you aren't sure which columns may have missing data in either of the two dataframes, so you'll need to do some work to find out.

Let's focus on looking at your ```parks``` dataframe first!

5. Check to see if the ```parks``` dataframe contains any missing values.  Check column names and data types as well
6. Take a look at the summary statistics
7. find out how many species observations there are for each National Park 

In [516]:
# 5

parks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23296 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  23296 non-null  object
 1   park_name        23296 non-null  object
 2   observations     23296 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 546.1+ KB


In [517]:
# 6

parks.describe(include='all')

scientific_name                            park_name  observations
count              23296                                23296  23296.000000
unique              5541                                    4           NaN
top     Myotis lucifugus  Great Smoky Mountains National Park           NaN
freq                  12                                 5824           NaN
mean                 NaN                                  NaN    142.287904
std                  NaN                                  NaN     69.890532
min                  NaN                                  NaN      9.000000
25%                  NaN                                  NaN     86.000000
50%                  NaN                                  NaN    124.000000
75%                  NaN                                  NaN    195.000000
max                  NaN                                  NaN    321.000000

In [518]:
# 7

parks.park_name.value_counts()

Great Smoky Mountains National Park    5824
Yosemite National Park                 5824
Bryce National Park                    5824
Yellowstone National Park              5824
Name: park_name, dtype: int64

You can see that the length of the dataframe is 23,296 rows long and that it has 3 columns, each with 0 missing values.  The column names are clean and easy to work with and you can also see the data types of each column make sense.

Looking at the output from code block #6, you can see that there are a lot of unique species within the dataframe (5541), but only 4 diferent National Parks.    

Since there isn't a lot of unique values, you can easily use ```.value_counts()``` on the dataframe to show that there is 5824 observations of species in each National Park.  

Though our data looks seemingly nice and tidy, we could further inspect the ```scientific_name``` column to see if there are any strange values that are hidden within the column that could effect the analysis.  An example of a strange value could be an empty string ("") or a string that says ```"NULL"``` or ```"NaN"```.  These would technically count as values but they should instead be classified as a missing value.

8. Pull some random samples of the ```scientific_name``` column, look over the values, and check to see if there are any entries that don't belong
  

In [519]:
# 8

for num in range(10):
    print(f'Sample {num + 1}')
    parks_sample = parks.scientific_name.sample(25)
    print(parks_sample)
    print()

Sample 1
22394                  Cardamine oligosperma
14914             Eriogonum nudum var. nudum
7518                       Lycopus uniflorus
13896                      Solidago puberula
7708                        Calidris bairdii
16881        Arabis laevigata var. laevigata
17721                        Liatris spicata
9241                 Polycarpon tetraphyllum
20219                     Caltha leptosepala
10235                        Ulota coarctata
3924                   Pseudotriton montanus
8678                     Osmunda claytoniana
13585                      Ipomoea pandurata
13618                   Lithophragma glabrum
7904     Phacelia ramosissima var. latifolia
8268                        Atrichum crispum
1056                         Veratrum viride
7721                       Salix ligulifolia
13440     Layia pentachaeta ssp. pentachaeta
10670     Crataegus douglasii var. douglasii
17965                      Amorpha fruticosa
13506             Saxifraga micranthidifolia
5

According to the samples, the data is good to go!

Time to check out the ```species``` dataframe!

9. Check to see if the ```species``` dataframe contains any missing values.  Check column names and data types as well
10. Take a look at the summary statistics
11. Find out how many species there are within each ```category```
12. Find out how many species there are within each ```conservation_status```
13. Find out the number of ```NaN``` values in the ```conservation_status``` column, and the percentage

In [520]:
# 9 

species.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5824 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5824 non-null   object
 1   scientific_name      5824 non-null   object
 2   common_names         5824 non-null   object
 3   conservation_status  191 non-null    object
dtypes: object(4)
memory usage: 182.1+ KB


In [521]:
# 10

species.describe(include='all')

category    scientific_name        common_names  \
count             5824               5824                5824   
unique               7               5541                5504   
top     Vascular Plant  Castor canadensis  Brachythecium Moss   
freq              4470                  3                   7   

       conservation_status  
count                  191  
unique                   4  
top     Species of Concern  
freq                   161

In [522]:
# 11

species.category.value_counts()

Vascular Plant       4470
Bird                  521
Nonvascular Plant     333
Mammal                214
Fish                  127
Amphibian              80
Reptile                79
Name: category, dtype: int64

In [523]:
# 12

species.conservation_status.value_counts()

Species of Concern    161
Endangered             16
Threatened             10
In Recovery             4
Name: conservation_status, dtype: int64

In [524]:
# 13

nan_count = len(species[species.conservation_status.isnull() == True])
nan_percent = (len(species[species.conservation_status.isnull() == True]) / len(species)) * 100

print(f'NaN values in \'conservation_status\' column: {nan_count} out of {len(species)}')
print(f'Percentage of NaN values in \'conservation_status\' column: {round(nan_percent, 2)}%')

NaN values in 'conservation_status' column: 5633 out of 5824
Percentage of NaN values in 'conservation_status' column: 96.72%


Going through the summary statistics gives us plenty of insight into the data.  The ```conservation_status``` column is clearly missing a massive chunk of data,
96.72% actually.  But why is this data missing?  If you take a look at the unique values within the column, you can see that there are only 4 values, and all of them are associated with endangerment to some degree (output of code block #12).  With this knowledge, we can assume that the ```NaN``` values just mean that the species is not at risk of endangerment.  The values seem to be missing for a reason.

With this assumption, you can go ahead and fill the missing values with a value that signifies "no risk".

14. Fill in the ```NaN``` values in the ```conservation_status``` column with the string value ```"Not at risk"```

In [525]:
# 14

species.conservation_status.fillna('Not at Risk', inplace=True)
species.conservation_status.value_counts()

Not at Risk           5633
Species of Concern     161
Endangered              16
Threatened              10
In Recovery              4
Name: conservation_status, dtype: int64

After doing this, it wouldn't hurt to check the scientific name column for any duplicate values.  If you refer back to code block #10, you can see that the column does have some repeating values.  This should be looked into.

15. Figure out how many values are duplicates within the ```scientific_name``` column
16. Find out how many duplicate rows exist within the ```species``` dataframe, if any

In [526]:
# 15

dup = species.scientific_name.duplicated().sum()
print(f'Duplicated values in \'scientific_name\' column: {dup}')

Duplicated values in 'scientific_name' column: 283


In [527]:
# 16 

dup = species.duplicated().sum()
print(f'Duplicated rows in \'species\' dataframe: {dup}')

Duplicated rows in 'species' dataframe: 0


So there aren't any duplicate rows within the ```species``` dataframe, but there are 283 duplicates within the ```scientific_name``` column of the ```species``` dataframe.  This may be due to the fact that more than one person recorded the same species and they recorded different numbers observed in the ```observations``` column.  Let's find out.

17. Experiment with the ```species``` dataframe and determine if these duplicate values should be dropped or not

In [528]:
# 17

species[species.scientific_name.duplicated()].head()

category         scientific_name                          common_names  \
3017   Mammal          Cervus elaphus                    Rocky Mountain Elk   
3019   Mammal  Odocoileus virginianus  White-Tailed Deer, White-Tailed Deer   
3020   Mammal             Canis lupus                       Gray Wolf, Wolf   
3022   Mammal           Puma concolor           Cougar, Mountain Lion, Puma   
3025   Mammal        Lutra canadensis                           River Otter   

     conservation_status  
3017         Not at Risk  
3019         Not at Risk  
3020         In Recovery  
3022         Not at Risk  
3025         Not at Risk

In [529]:
# 17 (Cont.)

species[species.scientific_name == 'Canis lupus']

category scientific_name     common_names conservation_status
8      Mammal     Canis lupus        Gray Wolf          Endangered
3020   Mammal     Canis lupus  Gray Wolf, Wolf         In Recovery
4448   Mammal     Canis lupus  Gray Wolf, Wolf          Endangered

This doesn't seem right.  This is the same mammal, recorded 3 times, and once with a different conservation status.  The only sensible explanation for the repeated entries could be due to a possible change in conservation status.  If that is the case, then we can rest knowing it was recorded for a reason. 

In [530]:
# 17 (Cont.)

species[species.scientific_name == 'Cervus elaphus']

category scientific_name        common_names conservation_status
4      Mammal  Cervus elaphus       Wapiti Or Elk         Not at Risk
3017   Mammal  Cervus elaphus  Rocky Mountain Elk         Not at Risk

In [531]:
# 17 (Cont.)

species[species.scientific_name == 'Puma concolor']

category scientific_name                 common_names conservation_status
16     Mammal   Puma concolor      Panther (Mountain Lion)         Not at Risk
3022   Mammal   Puma concolor  Cougar, Mountain Lion, Puma         Not at Risk
4451   Mammal   Puma concolor                Mountain Lion         Not at Risk

Ok, no rest for us, but you can conclude that a changing conservation status is NOT the reason for the duplicate values, and this is comforting, sort of.  It is safe to assume that the duplicates within the dataframe were recorded by different people who each had slightly different common names for the species.  To accurately plot any data, you'll need to drop the duplicates since they will do nothing for you but scew your results ever so slightly.

18. Drop duplicates in the ```scientific_name``` column of the ```species``` dataframe

In [532]:
species.drop_duplicates(subset=['scientific_name'], inplace=True)
dup = species.scientific_name.duplicated().sum()
print(f'Duplicated values in \'scientific_name\' column: {dup}')

Duplicated values in 'scientific_name' column: 0


18. Find out how many duplicate rows exist within the ```parks``` dataframe, if any
19. If so, drop the duplicated rows, you don't need any identical rows

In [533]:
# 18

parks.duplicated().sum()

15

In [534]:
# 19

print(f'Before dropping duplicates: {len(parks)}')
parks.drop_duplicates(inplace=True)
print(f'After dropping duplicates: {len(parks)}')

Before dropping duplicates: 23296
After dropping duplicates: 23281


20. The same ```scientific_name``` shouldn't be recorded at the same ```park_name``` more than once.  Find out how many duplicates you have when you check the duplicates between ```scientific_name``` and ```park_name```
21. There is a lot of duplicates.  Experiment with the ```parks``` dataframe and determine if these duplicate values should be dropped or not

In [535]:
# 20

dup = parks[['scientific_name', 'park_name']].duplicated().sum()
print(f'Duplicates: {dup}')

Duplicates: 1117


In [536]:
# 21

parks[parks[['scientific_name', 'park_name']].duplicated()]

scientific_name                  park_name  observations
483       Agrostis gigantea  Yellowstone National Park           235
490      Agrostis mertensii     Yosemite National Park           128
945           Rumex crispus  Yellowstone National Park           255
1213      Dianthus barbatus        Bryce National Park           110
1259        Riparia riparia        Bryce National Park            91
...                     ...                        ...           ...
23258  Microtus longicaudus  Yellowstone National Park           244
23259       Agrostis scabra        Bryce National Park           118
23267    Oxalis corniculata     Yosemite National Park           164
23273    Dactylis glomerata        Bryce National Park            89
23280    Botrychium simplex  Yellowstone National Park           241

[1117 rows x 3 columns]

In [537]:
# 21 (Cont.)

parks[parks.scientific_name == 'Agrostis gigantea']

scientific_name                            park_name  observations
449    Agrostis gigantea            Yellowstone National Park           253
483    Agrostis gigantea            Yellowstone National Park           235
6824   Agrostis gigantea                  Bryce National Park           104
7763   Agrostis gigantea  Great Smoky Mountains National Park            93
8676   Agrostis gigantea                  Bryce National Park           116
11602  Agrostis gigantea  Great Smoky Mountains National Park            57
13907  Agrostis gigantea               Yosemite National Park           148
17535  Agrostis gigantea               Yosemite National Park           128

In [538]:
# 21 (Cont.)

parks[parks.scientific_name == 'Oxalis corniculata']

scientific_name                            park_name  observations
415    Oxalis corniculata                  Bryce National Park            92
10750  Oxalis corniculata                  Bryce National Park            82
12378  Oxalis corniculata               Yosemite National Park           153
14294  Oxalis corniculata  Great Smoky Mountains National Park            91
18638  Oxalis corniculata            Yellowstone National Park           249
20252  Oxalis corniculata  Great Smoky Mountains National Park            73
21746  Oxalis corniculata            Yellowstone National Park           223
23267  Oxalis corniculata               Yosemite National Park           164

22. The duplicates should definitely be dropped, but which ones?  Some have higher ```observations``` values than others.  The best course of action would be to set the values within the observations column to the average number of observations for each duplicate ```scientific_name``` and ```park_name```.  Let's do just that

In [539]:
# 22

duplicated_name_and_park = parks[parks[['scientific_name', 'park_name']].duplicated()].reset_index(drop=True)

for i in range(len(duplicated_name_and_park)):
    
    scientific_name = duplicated_name_and_park.scientific_name.iloc[i]
    park_name = duplicated_name_and_park.park_name.iloc[i]

    duplicate = parks[(parks.scientific_name == scientific_name) & (parks.park_name == park_name)]
    avg_observation = int(duplicate.observations.mean())

    mask = parks[(parks.scientific_name == scientific_name) & (parks.park_name == park_name)]
    
    parks.loc[mask.index, 'observations'] = avg_observation

Let's check some duplicates to make sure our conversions worked

In [540]:
# 22 (Cont.)

parks[parks.duplicated()].head()

scientific_name                  park_name  observations
483    Agrostis gigantea  Yellowstone National Park           244
490   Agrostis mertensii     Yosemite National Park           131
945        Rumex crispus  Yellowstone National Park           255
1213   Dianthus barbatus        Bryce National Park            97
1259     Riparia riparia        Bryce National Park            89

In [541]:
# 22 (Cont.)

parks[(parks.scientific_name == 'Agrostis gigantea') & (parks.park_name == 'Yellowstone National Park')]

scientific_name                  park_name  observations
449  Agrostis gigantea  Yellowstone National Park           244
483  Agrostis gigantea  Yellowstone National Park           244

In [542]:
# 22 (Cont.)

parks[(parks.scientific_name == 'Riparia riparia') & (parks.park_name == 'Bryce National Park')]

scientific_name            park_name  observations
872   Riparia riparia  Bryce National Park            89
1259  Riparia riparia  Bryce National Park            89

Everything checks out!  The duplicate ```scientific_name``` and ```park_name``` values now have duplicate ```observations``` values as well, which are equal to the average of the observations between the duplicates.  Now we can simply just drop the duplicate rows of the dataframe since they are equal.

23. Now that the values have been correctly updated, you can drop the duplicates

In [543]:
# 23

print(f'Number of duplicates: {parks.duplicated().sum()}')

before = len(parks)
print(f'Length before dropping duplicates: {before}')

parks.drop_duplicates(inplace=True)

after = len(parks)
print(f'Length after dropping duplicates: {after}')

print(f'Difference between before and after: {before} - {after} = {before - after} duplicates dropped')
print(f'Duplicates in dataframe: {parks.duplicated().sum()}')

dup = parks[['scientific_name', 'park_name']].duplicated().sum()
print(f'Duplicates between \'scientific_name\' and \'park_name\' columns: {dup}')

Number of duplicates: 1117
Length before dropping duplicates: 23281
Length after dropping duplicates: 22164
Difference between before and after: 23281 - 22164 = 1117 duplicates dropped
Duplicates in dataframe: 0
Duplicates between 'scientific_name' and 'park_name' columns: 0


Perfect!  Now the ```parks``` dataframe contains each species number of observations ONCE per park.

The ```species``` and ```parks``` dataframes cleaned up and ready for further analysis!

24. Do a final summary statistics check on the ```species``` dataframe
25. Do a final summary statistics check on the ```parks``` dataframe

In [549]:
# 24

species.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5541 entries, 0 to 5823
Data columns (total 4 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   category             5541 non-null   object
 1   scientific_name      5541 non-null   object
 2   common_names         5541 non-null   object
 3   conservation_status  5541 non-null   object
dtypes: object(4)
memory usage: 216.4+ KB


In [550]:
# 24 (Cont.)

species.describe(include='all')

category                scientific_name        common_names  \
count             5541                           5541                5541   
unique               7                           5541                5229   
top     Vascular Plant  Clethrionomys gapperi gapperi  Brachythecium Moss   
freq              4262                              1                   7   

       conservation_status  
count                 5541  
unique                   5  
top            Not at Risk  
freq                  5363

In [551]:
# 25

parks.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 22164 entries, 0 to 23295
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   scientific_name  22164 non-null  object
 1   park_name        22164 non-null  object
 2   observations     22164 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 692.6+ KB


In [552]:
# 25 (Cont.)

parks.describe(include='all')

scientific_name                            park_name  observations
count                22164                                22164  22164.000000
unique                5541                                    4           NaN
top     Vicia benghalensis  Great Smoky Mountains National Park           NaN
freq                     4                                 5541           NaN
mean                   NaN                                  NaN    142.302472
std                    NaN                                  NaN     69.806198
min                    NaN                                  NaN      9.000000
25%                    NaN                                  NaN     86.000000
50%                    NaN                                  NaN    124.000000
75%                    NaN                                  NaN    195.000000
max                    NaN                                  NaN    321.000000

The dataframes look great!  Each dataframe contains 0 null values and 5541 unique values for the ```scientific_name``` column.  Which is the exact amount of species contained in the ```species``` dataframe.  The only time there will be a duplicate ```scientific_name``` value is when the scientific name is found in a different park!

# Appendix
### Section 1
1. Python library imports:
* Numpy - Used for fast calculations within pandas.  Can perform matrix calculations.
* Pandas - Used for loading in datasets into workable dataframes.  Crucial for Data Analysis with Python.
* Matplotlib - Used for creating plots and visualizations from pandas dataframes.
* Seaborn - Newer, more intuitive visualization library built to work with Matplotlib and pandas.
* Statsmodels - Used for creating predictive models.

2. Use the ```.read_csv()``` method along with ```pd``` (pandas) to import a csv file into a pandas dataframe.  The method takes the CSV filename surrounded by quotations as a parameter (Example: parks = pd.read_csv("observations.csv"))

3. Use the ```.head()``` method from the ```pandas``` library on the ```parks``` dataframe to show the first 10 rows of the dataframe

4. Use the ```.head()``` method from the ```pandas``` library on the ```species``` dataframe to show the first 10 rows of the dataframe

5. Use the ```.info()``` method from the ```pandas``` library on the ```parks``` dataframe to show column names, data types of the columns, length of the dataframe, and possible missing values within the dataframe.  This shows that the length of the dataframe is 23,296 rows long and that it has 3 columns, each with 0 missing values.  We can see that the column names are clean and easy to work with and we can also see the data types of each column, which are workable for now.

6. Use the ```.decribe()``` method from the ```pandas``` library on the ```parks``` dataframe to show some summary statistics about each column.  In order to see summary statistics for categorical variables, you need to use the ```include=``` parameter and set its value to ```'all'```.  This shows the amount of unique values for the categorical variables (```scientific_name``` and ```park_name```), and some useful statistics about the quantitative variable (```observations```).  The average number of times a single species was observed in 7 days was 142 times.  The minimum times observed was 9 times and the maximum was 321 times.  This is useful primarily because now you know that the ```observations``` column doesn't contain any zero values in it.  This means you can move on to exploring a different variable within the data.

7. Use the ```.value_counts()``` method from the ```pandas``` library on the ```parks.park_name``` dataframe column to show the frequency of the unique values within the column.  You can see that each of the 4 National Parks contain an equal amount of entries.  This information also shows that the 4 unique values within this column are in fact National Parks.

8. Create a ```for``` loop that loops 10 times.  With every loop print a string of the sample number.  With every loop you'll pull a random sample of the ```parks.scientific_name``` column using the pandas ```.sample()``` method.  The method will take an integer as a parameter which specifies the size of the random sample.  The larger the sample, the better.  After the samples are pulled, the samples should be examined for any values that are not scientific names.  If you pull enough samples of a decent size, then this can accurately portray the dataset as a whole.  This is much more efficient than reading through all 5541 unique scientific names.
